tensorflow/compiler/mlir/xla/transforms/lhlo_legalize_to_std.cc

```diff
@@ -0,0 +1,208 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements logic for lowering LHLO dialect to Affine dialect.
+
+#include "absl/memory/memory.h"
+#include "mlir/Dialect/AffineOps/AffineOps.h"  // TF:llvm-project
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // TF:llvm-project
+#include "mlir/IR/Attributes.h"  // TF:llvm-project
+#include "mlir/IR/Location.h"  // TF:llvm-project
+#include "mlir/IR/MLIRContext.h"  // TF:llvm-project
+#include "mlir/IR/PatternMatch.h"  // TF:llvm-project
+#include "mlir/IR/StandardTypes.h"  // TF:llvm-project
+#include "mlir/Transforms/DialectConversion.h"  // TF:llvm-project
+#include "mlir/Pass/Pass.h"  // TF:llvm-project
+#include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
+
+namespace mlir {
+namespace xla_lhlo {
+namespace {
+
+FlatSymbolRefAttr CallExternalFunc(
+    PatternRewriter &rewriter,
+    ModuleOp module,
+    Location loc,
+    const std::string& func_name,
+    std::vector<Type> &input_types,
+    std::vector<Type> &result_types,
+    ArrayRef<NamedAttribute> attrs) {
+
+  // TODO: FIXME, how about the same function name, by differernt params ?
+  // External function symbol is existed
+  FlatSymbolRefAttr func_name_attr = rewriter.getSymbolRefAttr(func_name);
+  if (module.lookupSymbol(func_name)) {
+    return func_name_attr;
+  }
+
+  auto context = rewriter.getContext();
+  // function type
+  auto func_type = FunctionType::get(input_types, result_types, context);
+
+  // create func op
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPointToStart(module.getBody());
+  rewriter.create<mlir::FuncOp>(loc, func_name_attr.getValue(),
+                                func_type, attrs);
+
+  return func_name_attr;
+}
+
+
+struct UniqueCountConverter : public OpRewritePattern<UniqueCountOp> {
```

Comment:
This is a lowering that would not be reusable on GPU or TPU, nor analyzable or fusable. This could be discussed separately. There is first the question of the HLOs being added and their semantics.